internal/controller/nab_init.go

```diff
@@ -0,0 +1,58 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+
+	nacv1alpha1 "github.com/migtools/oadp-non-admin/api/v1alpha1"
+	"github.com/migtools/oadp-non-admin/internal/common/constant"
+	"github.com/migtools/oadp-non-admin/internal/common/function"
+)
+
+// InitNonAdminBackup Initsets the New Phase on a NonAdminBackup object if it is not already set.
+//
+// Parameters:
+//
+//	ctx: Context for the request.
+//	log: Logger instance for logging messages.
+//	nab: Pointer to the NonAdminBackup object.
+//
+// The function checks if the Phase of the NonAdminBackup object is empty.
+// If it is empty, it sets the Phase to "New".
+// It then returns boolean values indicating whether the reconciliation loop should requeue
+// and whether the status was updated.
+func (r *NonAdminBackupReconciler) InitNonAdminBackup(ctx context.Context, log logr.Logger, nab *nacv1alpha1.NonAdminBackup) (exitReconcile bool, requeueReconcile bool, errorReconcile error) {
+	logger := log.WithValues("NonAdminBackup", nab.Namespace)
+	// Set initial Phase
+	if nab.Status.Phase == constant.EmptyString {
+		// Phase: New
+		updatedStatus, errUpdate := function.UpdateNonAdminPhase(ctx, r.Client, logger, nab, nacv1alpha1.NonAdminBackupPhaseNew)
+		if updatedStatus {
+			logger.V(1).Info("NonAdminBackup CR - Rqueue after Phase Update")
+			return false, true, nil
+		}
+		if errUpdate != nil {
```

Comment:
should not error be checked first than first return argument :question:

Reply:
no as in other comments we have 3 states here:
 - correct
 - correct with requeue (update status)
 - error

Reply:
These are just my two cents. I think you are saying that the states are mutually exclusive. There is only ever,

false, nil (correct)
true, nil (requeue)
false, err (error)

If that is the case, I would suggest that handling the third case first will look more go idiomatic for the future.

If there is a fourth option:
true, err 

If this is the case, an error will be required, but the status will not be updated IIUC. I think it makes sense if this is an important case, then we handle the request first so that we update the status.

Reply:
Updated to first check for errors. This was more of a 3rd state, which I checked can be safely moved to first check for the error. Closing as resolved as this is reworked in new commit. Please re-open if new implementation is also incorrect in your opinion.